api/envoy/config/listener/v3/listener.proto

```diff
@@ -109,6 +109,13 @@ message Listener {
   // Required unless *api_listener* or *listener_specifier* is populated.
   core.v3.Address address = 2;
 
+  // The additional addresses the listener should listen on. The address must be unqiue across all
+  // the listeners, also in the single listener. Multiple addresses with port 0 can be supported,
+  // same as the `address` field described. For multiple addresses in single listener,
+  // all addresses use the same protocol, and multiple internal addresses are not yet supported.
+  // [#not-implemented-hide:]
+  repeated core.v3.Address additional_addresses = 33;
```

Comment:
@adisuissa @ggreenway I give a try on `additional_addresses`. A little different with `one_of` way is `stat_prefix` has simple sematic. Another benefit is that will simple the code a little bit inside envoy, but I know it shouldn't be the reason for the API design, since the API is more for API user readable and expressive. Anyway, we can continue discussed here.

Original discussion link for other reviewers:
https://github.com/envoyproxy/envoy/pull/19367/files#r876516250

Reply:
Also seeing the discussion here https://github.com/envoyproxy/envoy/pull/21132#discussion_r864975841, that maybe the one of reason to use `one_of`. But I think udp and TCP should shared the address field, I don't get why should use `one_of` there.

Reply:
Would it be desirable to support listening on both a network address and as an `internal_listener`? I can imagine cases where that would be useful. cc @lambdai .

Reply:
Just quick through the code, support both network address and internal address seems workable. I guess that I just need to find out all the different processes between a network address and internal address in the ListenerImpl and ListenerManagerImpl, then make sure them reference the right config for the specific address, and create the right socket.  For ConnectionHandlerImpl, it works with different addresses ActiveListener works today.

now sure how much change that will be added, but should we support them separately? since already a lot of changes. If those are small changes, I think I can append those features quickly after initial implementation.

Reply:
I'm not saying they need to be supported immediately, but I don't think we should structure the config so that it can't be supported. So I'd say instead of a repeated Address, it should be a repeated message that currently contains an address (maybe inside a oneof). Then we can add more later without changing anything existing.

Reply:
I got your point now, thanks! Sorry I didn't follow that in the beginning.

The internal address is supported inside `Address`

https://github.com/envoyproxy/envoy/blob/4bb5b58394b4e8b6ee54db09ec4690d390027901/api/envoy/config/core/v3/address.proto#L136-L149

The check for rejecting the internal address is done by the code check https://github.com/envoyproxy/envoy/pull/19367/files#diff-f7639f0659fc4c081810d0a87d2e9e1aa9dc1c1423ea2d683ae491c4479e95f1R354-R361

When we want to support the internal address and both network and internal address, we can just remove the code check.

Reply:
Yea, I remember I was confused by the same thing before :)